tfx/examples/cifar10/cifar10_utils_native_keras.py

```diff
@@ -0,0 +1,204 @@
+LABEL_KEY = 'label'
+
+def transformed_name(key):
+  return key + '_xf'
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _get_serve_tf_examples_fn(model, tf_transform_output):
+  """Returns a function that parses a serialized tf.Example."""
+
+  @tf.function
+  def serve_tf_examples_fn(serialized_tf_examples):
+    """Returns the output to be used in the serving signature."""
+    feature_spec = tf_transform_output.raw_feature_spec()
+    feature_spec.pop(LABEL_KEY)
+    parsed_features = tf.io.parse_example(serialized_tf_examples, feature_spec)
+
+    transformed_features = tf_transform_output.transform_raw_features(
+        parsed_features)
+    # TODO(b/148082271): Remove this line once TFT 0.22 is used.
+    transformed_features.pop(transformed_name(LABEL_KEY), None)
+
+    return model(transformed_features)
+
+  return serve_tf_examples_fn
+
+def _input_fn(file_pattern: List[Text],
+              tf_transform_output: tft.TFTransformOutput,
+              batch_size: int = 200) -> tf.data.Dataset:
+  """Generates features and label for tuning/training.
+
+  Args:
+    file_pattern: List of paths or patterns of input tfrecord files.
+    tf_transform_output: A TFTransformOutput.
+    batch_size: representing the number of consecutive elements of returned
+      dataset to combine in a single batch
+
+  Returns:
+    A dataset that contains (features, indices) tuple where features is a
+      dictionary of Tensors, and indices is a single Tensor of label indices.
+  """
+  transformed_feature_spec = (
+      tf_transform_output.transformed_feature_spec().copy())
+
+  dataset = tf.data.experimental.make_batched_features_dataset(
+      file_pattern=file_pattern,
+      batch_size=batch_size,
+      features=transformed_feature_spec,
+      reader=_gzip_reader_fn,
+      label_key=transformed_name(LABEL_KEY))
+
+  return dataset
+
+def _freeze_model_by_percentage(model: tf.keras.Model,
+                                percentage: float):
+  """Freeze part of the model based on specified percentage
+
+  Args:
+    model: The keras model need to be partially frozen
+    percentage: the percentage of layers to freeze
+  """
+  percentage = max(0.0, percentage)
+  percentage = min(1.0, percentage)
+  num_layers = len(model.layers)
+  num_layers_to_freeze = int(num_layers * percentage)
+  for idx, layer in enumerate(model.layers):
+    if idx >= num_layers_to_freeze:
+      break
+    layer.trainable = False
+
+def _build_keras_model() -> tf.keras.Model:
+  """Creates a MobileNet model pretrained on ImageNet for finetuning
+
+  Returns:
+    A Keras Model.
+  """
+  tf.keras.backend.set_image_data_format('channels_last')
+  base_model = tf.keras.applications.MobileNet(
+      input_shape=(224, 224, 3), include_top=False, weights='imagenet',
+      pooling='avg')
+
+  _freeze_model_by_percentage(base_model, 0.9)
```

Comment:
My understanding is that the best practice is to train for a bit using the new final layer before unfreezing any part of the model to ensure that the new weights do not pollute the existing model. Can we please do that here?

Reply:
I have tried that setting, i.e. train new layers for 6 epochs and unfreeze all layers for another 6 epochs. But it didn't show improvement... Should I still add it?

Reply:
its good to highlight best practices so yes. and maybe write a comment that this part is "optional".